pkg/kubelet/status/status_manager.go

```diff
@@ -60,6 +60,9 @@ type versionedPodStatus struct {
 	// at is the time at which the most recent status update was detected
 	at time.Time
 
+	// True if the status is generated at the end of SyncTerminatedPod, or after it is completed.
+	podIsFinished bool
```

Comment:
I might be misunderstanding something here, but is there any reason we need `podIsFinished` as part of the version pod status? Why is it needed as a property on the struct? Can we just in `canBeDeleted` directly if the phase is terminal and call the new pod worker method `PodIsFinished`?

Reply:
I think the best outcome would be that we no longer need PodIsFinished.  The status manager is downstream of the pod worker and should know when it's finished reliably (TerminatePod is called).  We could model it not as part of this struct but a separate map (`m.podTerminated`) but we would have to clean that up separately and so not a net win (by contrast podStatusNew doesn't have to be cleaned up).

Reply:
> Why can't we just adjust `canBeDeleted` to check if the phase is terminal and call the new pod worker method `PodIsFinished`?

Other than conceptual (as pointed out by Clayton) there is an issue with only using `PodIsFinished` from workers. The problem is that `PodIsFinished` only starts to answer after `completeTerminated` which often happens after the update triggered by `SyncTerminatedPod`. That means, that we would need to wait 10s for the next `syncBatch` to trigger the update when `podWorkers.PodIsFinished=True` (tested). 

Second, the fallback to `podWorkers.PodIsFinished` is legacy - before I merged in the Claytons branch it could be that the status update from SyncTerminatedPod is lost if the status channel was full.

Reply:
I see, that makes sense, so the issue is that we call `TerminatePod` inside `SyncTerminatedPod` (https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet.go#L2110). But it means that if `PodIsFinished` is checked inside `TerminatePod` it would return false (because `SyncTerminatedPod` did not complete yet). This is because we call `p.completeTerminated(podUID)` (which sets finished) only after  `SyncTerminatedPod` exits.

It sounds like in either case though, we need to some signal that the pod is finished and can be deleted in status manager and need to propagate that signal from `TerminatePod` to `canBeDeleted(...)` and as @smarterclayton mentioned, having map or something like that would require additional cleanup.

Reply:
I think as we continue to evolve status manager it will be clearer what it needs.  It really is a subordinate of pod worker, and the sync*Pod methods drive changes to the real state for phase behavior, while readiness checks drive the status manager directly because those are orthogonal to phase (ish).